// @whatwg-node/server/esm/uwebsockets.js
import { ServerAdapterRequestAbortSignal } from './utils.js';
export function isUWSResponse(res) {
    return !!res.onData;
}
export function getRequestFromUWSRequest({ req, res, fetchAPI }) {
    let body;
    const method = req.getMethod();
    if (method !== 'get' && method !== 'head') {
        body = new fetchAPI.ReadableStream({});
        const readable = body.readable;
        res.onAborted(() => {
            readable.push(null);
        });
        let multipleChunks = false;
        res.onData(function (ab, isLast) {
            const chunk = Buffer.from(ab, 0, ab.byteLength);
            if (!multipleChunks && isLast) {
                readable.push(chunk);
            }
            else {
                readable.push(Buffer.from(chunk));
            }
            if (isLast) {
                readable.push(null);
            }
            multipleChunks = true;
        });
    }
    const headers = new fetchAPI.Headers();
    req.forEach((key, value) => {
        headers.set(key, value);
    });
    let url = `http://localhost${req.getUrl()}`;
    const query = req.getQuery();
    if (query) {
        url += `?${query}`;
    }
    return new fetchAPI.Request(url, {
        method,
        headers,
        body: body,
        signal: new ServerAdapterRequestAbortSignal(),
    });
}
async function forwardResponseBodyToUWSResponse(uwsResponse, fetchResponse) {
    let resAborted = false;
    uwsResponse.onAborted(function () {
        resAborted = true;
    });
    for await (const chunk of fetchResponse.body) {
        if (resAborted) {
            return;
        }
        uwsResponse.cork(() => {
            uwsResponse.write(chunk);
        });
    }
    uwsResponse.cork(() => {
        uwsResponse.end();
    });
}
export function sendResponseToUwsOpts(uwsResponse, fetchResponse) {
    if (!fetchResponse) {
        uwsResponse.writeStatus('404 Not Found');
        uwsResponse.end();
        return;
    }
    const bufferOfRes = fetchResponse._buffer;
    uwsResponse.cork(() => {
        uwsResponse.writeStatus(`${fetchResponse.status} ${fetchResponse.statusText}`);
        for (const [key, value] of fetchResponse.headers) {
            // content-length causes an error with Node.js's fetch
            if (key !== 'content-length') {
                if (key === 'set-cookie') {
                    const setCookies = fetchResponse.headers.getSetCookie?.();
                    if (setCookies) {
                        for (const setCookie of setCookies) {
                            uwsResponse.writeHeader(key, setCookie);
                        }
                        continue;
                    }
                }
                uwsResponse.writeHeader(key, value);
            }
        }
        if (bufferOfRes) {
            uwsResponse.end(bufferOfRes);
        }
    });
    if (bufferOfRes) {
        return;
    }
    if (!fetchResponse.body) {
        uwsResponse.end();
        return;
    }
    return forwardResponseBodyToUWSResponse(uwsResponse, fetchResponse);
}
